### PR TITLE
Add project integration PR helper

### PR DESCRIPTION
**Pull request by hogo6002**
Sample PRs (external users):
In the contact list: https://github.com/google/oss-fuzz/pull/10422#issuecomment-1567849463
Previous contributed: https://github.com/google/oss-fuzz/pull/10419#issuecomment-1567789169
Unknown contributor: https://github.com/google/oss-fuzz/pull/10416#issuecomment-1567755392
Integrating new project: https://github.com/google/oss-fuzz/pull/10439#issuecomment-1572939010

Skip commenting for internal members

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jonathanmetzman) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: Change "Upload" to "Integrate"

File: infra/tools/pr_helper/action.yml
Comment (jonathanmetzman) [inline infra/tools/pr_helper/action.yml:7]: Maybe don't include a default for required fields so that there's an error thrown (i assume) if it's not provided.

File: infra/tools/pr_helper/Dockerfile
Comment (jonathanmetzman) [inline infra/tools/pr_helper/Dockerfile:1]: nit: I think this file needs a license header.

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:89]: nit: single quotes.
Note that we use double quotes for docstrings.

**Comment by jonathanmetzman** [comment]
You should set up presubmit.py, then run format.
```
python3 -m venv .venv
source .venv/bin/activate
pip install -r infra/ci/requirements.txt
python infra/presubmit.py format
```

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:71]: nit: change to `get_pull_request_url`

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:83]: Hmm...I guess this works, but could you just create a string and write it to the file using `fh.write`.
There's nothing wrong with what you have here, but I haven't seen this in our code before and would rather keep things consistent.

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:92]: nit: just use an f-string.

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:104]: These shouldn't be f-strings since theres nothing we're formating in them.

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:114]: shouldn't be an fstring

File: infra/tools/pr_helper/helper.py
Comment (jonathanmetzman) [inline infra/tools/pr_helper/helper.py:123]: shouldn't be an f-string.

**Review by jonathanmetzman — APPROVED** [review]
Looks awesome. I don't think this needs another review from me before it lands. (as long as my comments are addressed). Maybe land this with @Alan32Liu 's help so he can revert if it causes problems.

**Review by Alan32Liu — COMMENTED** [review]
Looks pretty nice to me!
I am not 100% about the exact scope of the PR helper, so I took the chance to ask some questions.
Hope they make sense!

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (hogo6002) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: Yes, you are right. I am still adding more functions to it.

File: infra/tools/pr_helper/Dockerfile
Comment (hogo6002) [reply infra/tools/pr_helper/Dockerfile]: Removed docker

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (Alan32Liu) [reply .github/PULL_REQUEST_TEMPLATE.md]: Cool!
GitHub supports multiple PR templates, which could be suitable for our case.

Comment (hogo6002) [reply .github/PULL_REQUEST_TEMPLATE.md]: Removed PR templates, added criticality score in the comments

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:134]: Let's instead source this from a file in our repo, so it's more obvious to us where this is. e.g. a  MAINTAINERS file in the oss-fuzz repo root.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:43]: nit: can we do this in a more reliable way by checking that this is the top level `projects` directory in the oss-fuzz repo?

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:172]: nit: rather than pass headers around everywhere, let's wrap everything that talks to GitHub in a class and store the headers as a member variable there. 

Alternatively, we could also use something like https://github.com/PyGithub/PyGithub instead.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:55]: Can we check that this is verified by github? Commit authors can be forged easily, and we need to make sure this email is coming from a verified email on the github account.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:90]: nit: Can you put this in a CONSTANT global in this file? Alternatively, if /home/runner/go/bin is in PATH, use shutil.which to find the right path.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:185]: I think we should still add a message even if the repo_url is not set. Just say it's missing and can't compute criticality score.

**Review by oliverchang — COMMENTED** [review]
very nice!!

**Review by oliverchang — COMMENTED** [review]
very nice!!

File: infra/pr_helper.py
Comment (hogo6002) [reply infra/pr_helper.py:55]: I tried to use [Github email api](https://docs.github.com/en/rest/users/emails?apiVersion=2022-11-28#list-email-addresses-for-the-authenticated-user) first, but it can only obtain user's public email. For those who set up their email as private on github account (like myself), we can't obtain their verified emails by this way. So I used commit information instead. Please let me know if there is other approaches.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:213]: nit: this is a side effect that seems surprising based on the function name, which is just returning a boolean. We should move this logic to the caller.

File: infra/pr_helper.py
Comment (oliverchang) [reply infra/pr_helper.py]: There is also this "verification" field in the response (https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#get-a-commit) that we could leverage, but t his may not be set in all cases. 

If we can't verify either the email from the email API, nor from the verification field, then we should consider the email to be completely untrusted and forgeable and add this to our PR message. 

The other cases (checking for maintainer, or previous contributor), can be done by GitHub username right? Those we can continue to trust.

Comment (hogo6002) [reply infra/pr_helper.py]: Yes, the maintainer and previous contributor checks can be done by username. The email is only use to compare with the contact emails on project.yaml. Added `verification` field check. If the email is not verified, it will leave a message and continue previous contributor check.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:77]: nit: just do 

```python
verified, email = ...
```

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:180]: nit: return `False, None` here

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:99]: Should this be [1] here?

File: .github/workflows/pr_helper.yml
Comment (Alan32Liu) [inline .github/workflows/pr_helper.yml:70]: nit: As a non-native speaker, `Ready to merge` feels more natural to me :)

File: infra/pr_helper.py
Comment (Alan32Liu) [inline infra/pr_helper.py:32]: I reckon `os.path.join()` is normally preferred, but this should be OK given we always run this script on GitHub?

File: infra/pr_helper.py
Comment (Alan32Liu) [inline infra/pr_helper.py:73]: nit: Since we don't need variables like `messages` for this early return, I reckon it would be clearer to have:
```python3
  github = GithubHandler()
  # Bypasses PRs of the internal members.
  if github.is_author_internal_member():
    save_env(None, None, True)
    return

  message = ''
  is_ready_for_merge = True
  pr_author = github.get_pr_author()
  # Gets all modified projects path.
  projects_path = github.get_projects_path()
  email_response = github.get_author_email()

File: infra/pr_helper.py
Comment (Alan32Liu) [inline infra/pr_helper.py:83]: nit: period.

File: infra/pr_helper.py
Comment (Alan32Liu) [inline infra/pr_helper.py:95]:
```
if repo_url is None:
  ...
else:
  ...
```
Since we have both `if` and `else` now, I would suggest flipping their order to make the code easier to understand : )

File: infra/pr_helper.py
Comment (Alan32Liu) [inline infra/pr_helper.py:175]: Is this `if` block implying that `email` might not exist in `user_response.json()`?
If so, use `get('email')` to avoid `KeyError`.

**Review by Alan32Liu — APPROVED** [review]

File: infra/pr_helper.py
Comment (hogo6002) [reply infra/pr_helper.py:175]: The `email` will exist as response, but it may be `None`. The check is to make sure that it's not `None`

Comment (Alan32Liu) [reply infra/pr_helper.py:175]: Ah thanks! Then the convention is to check `None` values explicitly: e.g., `if email is None`.

File: infra/pr_helper.py
Comment (hogo6002) [reply infra/pr_helper.py:32]: i think it's OK as we only run on ubuntu machines. But please let me know if we have style requirements for path.

File: infra/pr_helper.py
Comment (hogo6002) [reply infra/pr_helper.py:175]: But I want to make sure that the string is not `None` or `empty`.

File: infra/pr_helper.py
Comment (Alan32Liu) [reply infra/pr_helper.py:32]: Yep, I am interested to learn that, too.

Comment (oliverchang) [reply infra/pr_helper.py:32]: Usually yes we would prefer os.path.oin. In this case though, even if we do os.path.join it's rather hardcoded to the Ubuntu GitHub actions runner so it would not be adding much value.

File: infra/pr_helper.py
Comment (oliverchang) [inline infra/pr_helper.py:99]: nit: prefer just "if email" per https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations

**Review by oliverchang — APPROVED** [review]

**Review by oliverchang — APPROVED** [review]
nice!!

File: infra/pr_helper.py
Comment (Alan32Liu) [reply infra/pr_helper.py]: Ah sorry, I thought we prefer [checking `None` explicitly](https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations:~:text=Always%20use%20if%20foo%20is%20None%3A)?

Comment (oliverchang) [reply infra/pr_helper.py]: That's only needed if we need to explicitly check between a None value vs a an empty (e.g. '') value. In this case, it makes no difference.

Comment (hogo6002) [reply infra/pr_helper.py]: oh, I see the difference now. Thanks

Comment (Alan32Liu) [reply infra/pr_helper.py]: I see, thanks!

**Pull request merged**
